### PR TITLE
Template args

### PR DESCRIPTION
**Pull request by Slava**
A WIP PR and discussion for the template arguments. This branch is based on the `each-in` branch and `spacebars-template-exprs` branch.

```handlebars
<template name="myTemplate" args="fistName lastName">
  <p>{{firstName}} - {{lastName}}</p>
  <p>All arguments:</p>
  <ul>
    {{#each val in @args}}
        <li>{{@key}}: {{val}}</li>
    {{/each}}
  </ul>
</template>

<template name="main">
  {{@> myTemplate firstName="John" lastName="Doe" age=34 nationality="U.S."}}
</template>
```

Will output:

```html
<p>John - Doe</p>
<p>All arguments</p>
<ul>
  <li>firstName: John</li>
  <li>lastName: Doe</li>
  <li>age: 34</li>
  <li>nationality: U.S.</li>
</ul>
```

What's new:

1. The `@>` operator is the new inclusion with args (as opposed to `>` which is inclusion with the data context). Positional and keyword args are under consideration.
2. `@args` - special keyword to access a dictionary of passed arguments. Keyword args are mapped key-value, positional args are mapped as 0-arg0, 1-arg1. Iterable with `#each`.
3. Template's args signature: defined args will be dumped into scope for matching passed args. Other args, not present in the signature, will be available only in `@args`.
4. `@args` can be accessed in JS helpers as `Template.instance().args.get('x')` to get argument `x`. Similar to `Template.instance().state.get('x')/set('x', y)`.

### CODE REVIEW

**Comment by stubailo** [comment]
@Slava You should make this PR to the other branch with scope stuff, it's hard to tell which parts are actually related to args

File: packages/blaze/lookup.js [hunk 1]
```diff
@@ -105,38 +134,52 @@ Blaze.View.prototype.lookup = function (name, _options) {
...
+
+  // 1. look up a helper on the current template
+  if (template && ((helper = getTemplateHelper(template, name)) != null)) {
     return wrapHelper(bindDataContext(helper), boundTmplInstance);
-  } else if (lookupTemplate && (name in Blaze.Template) &&
+  }
+
+  // 2. look up a binding by traversing the lexical view hierarchy inside the
+  // current template
+  if (template && (binding = lexicalBindingLookup(Blaze.currentView, name)) != null) {
```
Comment (mitar) [inline packages/blaze/lookup.js]: Isn't `this === Blaze.currentView`?

File: packages/blaze/lookup.js [hunk 2]
```diff
@@ -73,6 +73,33 @@ var wrapHelper = function (f, templateFunc) {
   };
 };
 
+var lexicalBindingLookup = function (view, name) {
+  var currentView = view;
+  var blockHelpersStack = [];
```
Comment (mitar) [inline packages/blaze/lookup.js]: This variable is unused.

**Comment by ryneeverett** [comment]
What ended up happening with this? It's in a different branch now?

**Pull request closed without merging**